text/0000-simple_unwind_annotation.md

```diff
@@ -0,0 +1,78 @@
+- Feature Name: `simple_unwind_attribute`
+- Start Date: 2019-08-29
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provides an annotation to permit functions with explicit ABI specifications
+(such as `extern "C"`) to unwind
+
+# Motivation
+[motivation]: #motivation
+
+TODO
+
+- soundness & optimization
+- libjpeg/mozjpeg
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `#[unwind(allowed)]` attribute permits functions with non-Rust ABIs (e.g. `extern "C" fn`) to unwind rather than terminating the process.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+By default, Rust assumes that an external function imported with `extern "C" {
+... }` (or another ABI other than `"Rust"`) cannot unwind, and Rust will abort if a panic would propagate out of a
+Rust function with a non-"Rust" ABI ("`extern "ABI" fn`") specification. If you specify
+the `#[unwind(allowed)]` attribute on a function with a non-"Rust" ABI, Rust
+will instead allow an unwind (such as a panic) to proceed through that function
+boundary using Rust's normal unwind mechanism. This may potentially allow Rust
+code to call non-Rust code that calls back into Rust code, and then allow a
+panic to propagate from Rust to Rust across the non-Rust code.
+
+The Rust unwind mechanism is intentionally not specified here. Catching a Rust
+panic in Rust code compiled with a different Rust toolchain or options is not
+specified. Catching a Rust panic in another language or vice versa is not
+specified. The Rust unwind may or may not run non-Rust destructors as it
+unwinds. Propagating a Rust panic through non-Rust code is unspecified;
+implementations that define the behavior may require target-specific options
+for the non-Rust code, or this feature may not be supported at all.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Only works as long as Rust uses the same unwinding mechanism as C++.
+- Does not allow external library bindings to specify whether callbacks they accept are expected to unwind.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
```

Comment:
> So I'd like to know which alternatives are there, what solutions did people actually try, with code, etc. so that we can understand why didn't they work out. Because AFAICT, no single stakeholder in this matter has actually tried to write any code to solve this problem.

We have invested quite a bit of time in Lucet and Cranelift to get unwinding to work across generated code. The PRs are on hold awaiting stabilization of `#[unwind]`, but if you're interested you can see them here: ([Lucet](https://github.com/fastly/lucet/pull/254), [Cranelift](https://github.com/CraneStation/cranelift/pull/902)). I don't think it's helpful to approach this issue as though others have not put quite a bit of thought and effort into addressing it.

Like many engineering challenges, this is less about other solutions not working out than about trying to avoid undesirable tradeoffs. We don't want the overhead of a `setjmp` every time we enter generated code, nor do we want the extra complexity and bookkeeping of maintaining a stack of jump buffers. We also don't want to sacrifice late binding of WebAssembly imports in order to fit into C wrappers, or to incur special-case complexity to change our compiler to generate the wrappers in the Lucet shared objects.

Yes, depending on implementation-defined behavior for the unwinding mechanism instead comes with its own drawbacks, but these are mitigated by the fact that we are targeting a specific platform and have control over both our Rust compiler and our code generator. The one piece that's out of our control now is that stable Rust insists on emitting these `nounwind`s.